### PR TITLE
catch errors and retry. This should handle connection resets during long syncs

### PR DESCRIPTION
**Pull request by improbable-mattchurch**
See: https://stackoverflow.com/questions/51381311/p4-python-connection-broken-ssl-error

### CODE REVIEW

File: python/perforce.py [hunk 1]
```diff
@@ -132,6 +132,22 @@ def _write_patched(self, files):
         with open(self.patchfile, 'w') as outfile:
             json.dump(content, outfile)
 
+    # decorator for reconnecting to perforce and retrying a function on P4Exception
+    def reconnect_on_exception(self, f, retries=3):
+        def wraps(*args):
+            success = False
+            while(retries > 0 and not success):
+                try:
+                    f(args)
```
Comment (petemounce) [inline python/perforce.py]:
```suggestion
                    f(args)
                    success = True
```

**Review by petemounce — COMMENTED** [review]
Would it be reasonable to add some test coverage?

**Comment by improbable-mattchurch** [comment]
It would be difficult to test this properly without mocking a p4 server and dropping the client's connection

**Comment by mattyoung-improbable** [comment]
These connection resets are timeouts I think? So this'll retry after like 2 hours?

**Comment by improbable-mattchurch** [comment]
closing as https://github.com/improbable-eng/perforce-buildkite-plugin/pull/172 likely will solve our issue

**Pull request closed without merging**